File: backend/src/main/java/moe/mewore/rabbit/backend/physics/RigidBodyController.java
package moe.mewore.rabbit.backend.physics;

import javax.vecmath.Vector2f;
import javax.vecmath.Vector3f;
import java.util.function.Consumer;

import com.bulletphysics.collision.broadphase.Dispatcher;
import com.bulletphysics.collision.dispatch.CollisionWorld;
import com.bulletphysics.collision.narrowphase.PersistentManifold;
import com.bulletphysics.dynamics.ActionInterface;
import com.bulletphysics.dynamics.RigidBody;
import com.bulletphysics.linearmath.IDebugDraw;
import com.bulletphysics.linearmath.Transform;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class RigidBodyController extends ActionInterface {

    private static final float JUMP_SPEED = 110f;

    private static final float MAX_Y_SPEED = JUMP_SPEED * 2f;

    private static final float ACCELERATION = 400f;

    private static final float GROUND_CHECK_REQUIRED_NORMAL_Y = .3f;

    private final RigidBody body;

    private final Transform tmpTransform = new Transform();

    private final Vector3f tmpVector3 = new Vector3f();

    public float groundLeniency = 0.1f;

    public float jumpControlLeniency = 0.1f;

    public Vector2f targetHorizontalMotion = new Vector2f();

    public float groundTimeLeft = -1f;

    public float jumpControlTimeLeft = -1f;

    public Vector3f getPosition() {
        return body.getWorldTransform(tmpTransform).origin;
    }

    public void setPosition(final Vector3f position) {
        tmpTransform.setIdentity();
        tmpTransform.origin.set(position);
        body.setWorldTransform(tmpTransform);
    }

    public void loadPosition(final float[] data, int index) {
        tmpTransform.setIdentity();
        tmpTransform.origin.set(data[index], data[++index], data[++index]);
        body.setWorldTransform(tmpTransform);
    }

    public Vector3f getMotion(final Vector3f target) {
        return body.getLinearVelocity(target);
    }

    public Vector3f getMotion() {
        return getMotion(tmpVector3);
    }

    public void setMotion(final Vector3f motion) {
        body.setLinearVelocity(motion);
    }

    public void loadMotion(final float[] data, int index) {
        tmpVector3.set(data[index], data[++index], data[++index]);
        body.setLinearVelocity(tmpVector3);
    }

    public void jump() {
        jumpControlTimeLeft = jumpControlLeniency;
    }

    boolean onGround() {
        return groundTimeLeft >= 0;
    }

    private boolean wantsToJump() {
        return jumpControlTimeLeft >= 0;
    }

    @Override
    public void updateAction(final CollisionWorld world, final float deltaTimeStep) {
        final var motion = getMotion(tmpVector3);
        final var motionDx = targetHorizontalMotion.x - motion.x;
        final var motionDz = targetHorizontalMotion.y - motion.z;
        final var motionDistanceSquared = motionDx * motionDx + motionDz * motionDz;
        final var maxFrameAcceleration = ACCELERATION * deltaTimeStep;
        if (motionDistanceSquared > maxFrameAcceleration * maxFrameAcceleration) {
            final var multiplier = maxFrameAcceleration / Math.sqrt(motionDistanceSquared);
            motion.x = (float) (motion.x + motionDx * multiplier);
            motion.z = (float) (motion.z + motionDz * multiplier);
        } else {
            motion.x = targetHorizontalMotion.x;
            motion.z = targetHorizontalMotion.y;
        }
        if (onGround() && wantsToJump()) {
            motion.y = JUMP_SPEED;
            groundTimeLeft = jumpControlTimeLeft = 0;
        } else {
            motion.y = Math.max(-MAX_Y_SPEED, Math.min(MAX_Y_SPEED, motion.y));
        }

        body.setLinearVelocity(motion);

        if (Math.abs(motion.x) > 0 || Math.abs(motion.y) > 0 || Math.abs(motion.z) > 0) {
            body.activate();
        }

        jumpControlTimeLeft -= deltaTimeStep;
        groundTimeLeft -= deltaTimeStep;
    }

    @Override
    public void debugDraw(final IDebugDraw debugDrawer) {
        // Do nothing for now
    }

    public void afterPhysics(final CollisionWorld world) {
        forEachGroundCollision(world, collisionNormal -> groundTimeLeft = groundLeniency);
    }

    private void forEachGroundCollision(final CollisionWorld world, final Consumer<Vector3f> callback) {
        final Dispatcher dispatcher = world.getDispatcher();
        for (int i = 0; i < dispatcher.getNumManifolds(); i++) {
            final PersistentManifold manifold = dispatcher.getManifoldByIndexInternal(i);
            final var isFirstBody = manifold.getBody0() == body;
            final var isSecondBody = !isFirstBody && manifold.getBody1() == body;
            if (isFirstBody || isSecondBody) {
                for (int j = 0; j < manifold.getNumContacts(); j++) {
                    final var normal = manifold.getContactPoint(j).normalWorldOnB;
                    if ((isFirstBody ? (1 - normal.y) : (1 + normal.y)) < GROUND_CHECK_REQUIRED_NORMAL_Y) {
                        tmpVector3.set(normal);
                        if (!isFirstBody) {
                            tmpVector3.negate();
                        }
                        callback.accept(tmpVector3);
                    }
                }
            }
        }
    }
}

File: frontend/src/game/physics/rigid-body-controller.ts
import Ammo from 'ammo.js';
import { btCollisionWorld } from 'ammo.js';

const ACCELERATION = 400;

export const JUMP_SPEED = 110;

export const MAX_Y_SPEED = JUMP_SPEED * 2;

const GROUND_CHECK_REQUIRED_NORMAL_Y = 0.3;

type Vector3 = { readonly x: number; readonly y: number };

let tmpVector3: Ammo.btVector3;

export class RigidBodyController {
    groundLeniency = 0.1;
    jumpControlLeniency = 0.1;

    targetHorizontalMotion: Vector3 = { x: 0, y: 0 };

    groundTimeLeft = 0;
    jumpControlTimeLeft = 0;

    constructor(private readonly body: Ammo.btRigidBody) {}

    get position(): Ammo.btVector3 {
        return this.body.getWorldTransform().getOrigin();
    }

    get motion(): Ammo.btVector3 {
        return this.body.getLinearVelocity();
    }

    jump(): void {
        this.jumpControlTimeLeft = this.jumpControlLeniency;
    }

    onGround(): boolean {
        return this.groundTimeLeft >= 0;
    }

    private wantsToJump(): boolean {
        return this.jumpControlTimeLeft >= 0;
    }

    updateAction(_world: Ammo.btCollisionWorld, deltaTimeStep: number): void {
        const motion = this.body.getLinearVelocity();
        const motionDx = this.targetHorizontalMotion.x - motion.x();
        const motionDz = this.targetHorizontalMotion.y - motion.z();
        const motionDistanceSquared = motionDx * motionDx + motionDz * motionDz;
        const maxFrameAcceleration = ACCELERATION * deltaTimeStep;
        if (motionDistanceSquared > maxFrameAcceleration * maxFrameAcceleration) {
            const multiplier = maxFrameAcceleration / Math.sqrt(motionDistanceSquared);
            motion.setX(motion.x() + motionDx * multiplier);
            motion.setZ(motion.z() + motionDz * multiplier);
        } else {
            motion.setX(this.targetHorizontalMotion.x);
            motion.setZ(this.targetHorizontalMotion.y);
        }
        if (this.onGround() && this.wantsToJump()) {
            motion.setY(JUMP_SPEED);
            this.groundTimeLeft = this.jumpControlTimeLeft = 0;
        } else {
            motion.setY(Math.max(-MAX_Y_SPEED, Math.min(MAX_Y_SPEED, motion.y())));
        }

        this.body.setLinearVelocity(motion);

        if (Math.abs(motion.x()) > 0 || Math.abs(motion.y()) > 0 || Math.abs(motion.z()) > 0) {
            this.body.activate();
        }

        this.jumpControlTimeLeft -= deltaTimeStep;
        this.groundTimeLeft -= deltaTimeStep;
    }

    afterPhysics(world: Ammo.btCollisionWorld): void {
        this.forEachGroundCollision(world, () => (this.groundTimeLeft = this.groundLeniency));
    }

    private forEachGroundCollision(world: btCollisionWorld, callback: (normal: Ammo.btVector3) => void): void {
        const dispatcher = world.getDispatcher();
        for (let i = 0; i < dispatcher.getNumManifolds(); i++) {
            const manifold = dispatcher.getManifoldByIndexInternal(i);
            const isFirstBody = Ammo.castObject(manifold.getBody0(), Ammo.btRigidBody) === this.body;
            const isSecondBody = !isFirstBody && Ammo.castObject(manifold.getBody1(), Ammo.btRigidBody) === this.body;
            if (isFirstBody || isSecondBody) {
                for (let j = 0; j < manifold.getNumContacts(); j++) {
                    const normal = manifold.getContactPoint(j).get_m_normalWorldOnB();
                    if ((isFirstBody ? 1 - normal.y() : 1 + normal.y()) < GROUND_CHECK_REQUIRED_NORMAL_Y) {
                        if (isFirstBody) {
                            callback(normal);
                        } else {
                            const fixedNormal = (tmpVector3 = tmpVector3 || new Ammo.btVector3());
                            fixedNormal.setValue(-normal.x(), -normal.y(), -normal.z());
                            callback(fixedNormal);
                        }
                    }
                }
            }
        }
    }
}
